src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala

```diff
@@ -380,7 +380,31 @@ class PostgresDirectoryDAO(protected val dbRef: DbReference,
 
   override def readProxyGroup(userId: WorkbenchUserId): IO[Option[WorkbenchEmail]] = ???
 
-  override def listUsersGroups(userId: WorkbenchUserId): IO[Set[WorkbenchGroupIdentity]] = ???
+  override def listUsersGroups(userId: WorkbenchUserId): IO[Set[WorkbenchGroupIdentity]] = {
+    runInTransaction { implicit session =>
+      val ancestorGroupsTable = SubGroupMemberTable("ancestor_groups")
+      val ag = ancestorGroupsTable.syntax("ag")
+      val agColumn = ancestorGroupsTable.column
+
+      val gm = GroupMemberTable.syntax("gm")
+      val pg = GroupMemberTable.syntax("parent_groups")
+      val g = GroupTable.syntax("g")
+
+      val listGroupsQuery =
+        samsql"""WITH RECURSIVE ${ancestorGroupsTable.table}(${agColumn.parentGroupId}, ${agColumn.memberGroupId}) AS (
+                    select ${gm.groupId}, ${gm.memberGroupId}
+                    from ${GroupMemberTable as gm}
+                    where ${gm.memberUserId} = ${userId}
+                    union
+                    select ${pg.groupId}, ${pg.memberGroupId}
+                    from ${GroupMemberTable as pg}
+                    join ${ancestorGroupsTable as ag} ON ${agColumn.parentGroupId} = ${pg.memberGroupId}
+          ) select distinct(${g.name})
```

Comment:
unfortunately WorkbenchGroupIdentity is not only WorkbenchGroups but also policies... you need to include possible policies in the result set